### ISSUE TITLE
[Core] Support export of ray jobs and actors history to external storage 

### ISSUE THREAD

**Issue by jhasm-ck** [ray-project/ray#39503]
### Description

1. The ray jobs, actors, tasks history and all the information showed on the ray dashboard is currently limited to a running ray cluster and is lost when the cluster is deleted. 
2. The ray clusters are designed to be multi-tenant or long running, but designed to be created to run a few jobs and then destroyed.

The above two factors together mean the ray dashboard information and job history is very limited in use. All the work put in to improve the dashboard has limited usefulness. To use the full potential of the history and other dashboard information, they need to have a longer life than the cluster itself. We do support exporting logs and metrics through popular integrations, but there is no support for exporting job, actor and task history along with all the rich debugging information we have on the dashboard.

I suggest a mechanism to listen to events triggered from the ray cluster, so that the listener can store the history in a more durable storage of their choice. Pulling information from the cluster is not reliable because the clusters can go up and down.  A push mechanism is more reliable and resource efficient in this situation.

### Use case

1. To be able to see the history of jobs run by me or my team, independent of which ray cluster was used.
2. Being able to persist the job information without keeping the ray cluster running.
3. Being able to see the history of all jobs in one place, across all ray clusters.
4. Being able to treat the kuberay clusters as lightweight resources, without loosing the history of jobs and debugging information.
5. Being able to record the ray job information along with the rest of the model logs and artifacts.

**Comment by jhasm-ck** [comment]
cc @richardliaw

**Comment by scottsun94** [comment]
This is more related to persistent dashboard:  exporting the info from GCS that powers that dashboard.

There has been related discussion: @alanwguo @gvspraveen

**Comment by shrekris-anyscale** [comment]
cc @akshay-anyscale

**Comment by azai91** [comment]
Any update to this feature? We spin up/teardown Ray clusters frequently and would to be able to persist jobs/actor history.

**Comment by scottsun94** [comment]
cc: @anyscalesam

**Comment by askulkarni2** [comment]
This seems like a valid use-case. Has any thought been given to de-coupling the dashboard from Ray clusters? On kubernetes deployments a single cluster-wide dashboard which can display information from multiple RayCluster resources would make sense.

**Comment by kadisi** [comment]
Any update to this feature?

**Comment by scottsun94** [comment]
cc: @nikitavemuri @alanwguo

**Comment by xiaoming12306** [comment]
Any update to this feature?

**Comment by nikitavemuri** [comment]
Hi all, we've been actively developing a push-based solution for exporting state and metadata for various Ray resources. We have an alpha version of this feature for some core Ray resources, but we've observed some performance regressions in task scheduling latency and throughput. We are currently investigating these issues and exploring alternative architectures that may offer better performance. Expecting to provide an update on this work in Q1 next year.

**Comment by kadisi** [comment]
> Hi all, we've been actively developing a push-based solution for exporting state and metadata for various Ray resources. We have an alpha version of this feature for some core Ray resources, but we've observed some performance regressions in task scheduling latency and throughput. We are currently investigating these issues and exploring alternative architectures that may offer better performance. Expecting to provide an update on this work in Q1 next year.

hi @nikitavemuri , We built on existing ray dashboard capabilities, implemented ray historyserver internally by providing additional components, and persisted metadata and logs to shared storage.   Based on this ability，we can view multiple ray cluster dashboards online and offline. But we want raydashboard itself to support persistent metadata information. This will make raydashbord easier to integrate.  Is there any further planning and design for this area? We would be very interested in that

**Comment by anapivotal** [comment]
Hey! Is this feature still planned to implemented?

**Comment by nikitavemuri** [comment]
Yes, we are still actively working on this feature. We previously implemented an alpha version of the export API but observed regressions with large workloads, so this feature was redesigned to be more performant. We expect to release task state and metadata events this quarter, which can be used to persist the task table view, and will be following up with additional Ray resource types (eg: actors, nodes, jobs).

**Comment by jhasm** [comment]
Thanks for the update @nikitavemuri ! Is there a way we can see and track the progress on this work? It will help us plan our own work around this feature.

**Comment by weimingdiit** [comment]
Thanks for the update @nikitavemuri . Looking forward to the launch of this feature!

**Comment by metasyn** [comment]
Slack is interested in this workstream and feature as well. Maybe there will be something about it at the https://www.anyscale.com/ray-summit/2025 🤞

**Comment by MengjinYan** [comment]
Hi all, we just introduced the alpha release of the Ray Event Export API in the Ray Summit. This is a push-based mechanism that can send the Ray workload execution information to a pre-configured http endpoint from each nodes in the Ray cluster. In this way, the information can further be persisted for analysis after cluster termination. 

For more information, you can check out the Ray doc here: https://docs.ray.io/en/master/ray-observability/user-guides/ray-event-export.html 

We also have a community announcement here: https://github.com/ray-project/ray/issues/56057 

Let us know if you tried it out and encounter any issues or if you have any feedback to the event schema.
